src/dmqproto/client/request/internal/Consume.d

```diff
@@ -12,6 +12,711 @@
+            fiber,
+            (conn.Payload payload)
+            {
+                payload.addConstant(MessageType_v3.Stop);
+            }
+        );
+    }
+
+    /***************************************************************************
+
+        The fiber that waits for a batch of records to arrive and passes it to
+        the user, then sends the `Continue` message to the node, in a loop.
+        Handles suspending the request through the controller, for resuming call
+        `resume`. Calling `stop` makes this routine terminate after all
+        remaining records have been passed to the user.
+
+    ***************************************************************************/
+
+    private class RecordStream
+    {
+        /// The acquired buffer to store a batch of records.
+        private Const!(void)[]* batch_buffer;
+
+        /// Slices the records in *batch_buffer that haven't been processed yet.
+        private Const!(void)[] remaining_batch = null;
+
+        /// The fiber.
+        private MessageFiber fiber;
+
+        /// Tells if the fiber is suspended, and if yes, what it is waiting for.
+        enum FiberSuspended: uint
+        {
+            No,
+            WaitingForRecords,
+            RequestSuspended
+        }
+
+        private FiberSuspended fiber_suspended;
+
+        /// A Stopped message has been received.
+        private bool stopped = false;
+
+        /// Token passed to fiber suspend/resume calls.
+        private static MessageFiber.Token token =
+            MessageFiber.Token(typeof(this).stringof);
+
+        /// Constructor, starts the fiber.
+        private this ( )
+        {
+            this.batch_buffer = this.outer.resources.getBuffer();
+            this.fiber = this.outer.resources.getFiber(&this.fiberMethod);
+            this.fiber.start();
+        }
+
+        /***********************************************************************
+
+            Adds a batch of records to be passed to the user notifier and
+            resumes the fiber if it is waiting for more records. Callen when a
+            `Records` message from the node has arrived.
+
+            Params:
+                record_batch = the batch of records to add
+
+        ***********************************************************************/
+
+        public void addRecords ( in void[] record_batch )
+        {
+            // Append record_batch to *this.batch_buffer, which may or may not
+            // be empty.
+
+            if (!(*this.batch_buffer).length)
+                enableStomping(*this.batch_buffer);
+
+            // Append record_batch, then set this.remaining_batch to reference
+            // the remaining records. To avoid a dangling slice if
+            // *this.batch_buffer is relocated, set this.remaining_batch to null
+            // first.
+            size_t n_processed = (*this.batch_buffer).length - this.remaining_batch.length;
+            this.remaining_batch = null;
```

Comment:
I don't get why setting this slice to null matters here, as it's reset two lines down, anyway.

Reply:
It is so that `this.remaining_batch` cannot be a dangling slice, not even during the execution of one line of source code. I admit it's a bit nitpicking.